File: day3/task1.zig
const std = @import("std");

const Santa = struct {
  x: i32,
  y: i32,
  houses: u32,
  visited: std.AutoHashMap([2]i32, bool),
  const Self = @This();
  fn new(allocator: std.mem.Allocator) !Self {
    var visited = std.AutoHashMap([2]i32, bool).init(allocator);
    try visited.put([2]i32{0, 0}, true);
    return Self {
      .x = 0,
      .y = 0,
      .houses = 1,
      .visited = visited,
    };
  }
  inline fn deinit(self: *Self) void {
    self.visited.deinit();
  }
  fn move(self: *Self, direction: u8) !void {
    switch (direction) {
      '>' => { self.x += 1; },
      'v' => { self.y -= 1; },
      '<' => { self.x -= 1; },
      '^' => { self.y += 1; },
      // else => return error.UnknownDirection,
      else => {},
    }
    try self.check_house();
  }
  fn check_house(self: *Self) !void {
    const position = [2]i32{self.x, self.y};
    if (self.visited.contains(position)) {
      return;
    }
    try self.visited.put(position, true);
    self.houses += 1;
  }
};

pub fn main() !void {
  var file = try std.fs.cwd().openFile("input.txt", .{});
  defer file.close();
  var buffered = std.io.bufferedReader(file.reader());
  var reader = buffered.reader();
  var buffer: [100 * 1024]u8 = undefined;
  var fba = std.heap.FixedBufferAllocator.init(&buffer);
  const allocator = fba.allocator();
  var santa = try Santa.new(allocator);
  defer santa.deinit();
  while (true) {
    const symbol = reader.readByte() catch |err| switch (err) {
      error.EndOfStream => break,
      else => return err,
    };
    try santa.move(symbol);
  }
  std.debug.print("Houses: {}", .{santa.houses});
}

File: day3/task1.pas
{$mode objfpc}

type
  pair = array[0..1] of integer;
  positions = array of pair;
  positions_ref = ^positions;
  santa = class
    private
      position: pair;
      visited: positions_ref;
    public
      constructor init (
        const init_position: pair; 
        const init_visited: positions_ref
      );
      procedure follow(const direction: char);
  end;

constructor santa.init (
  const init_position: pair; 
  const init_visited: positions_ref
); 
begin
  self.position := init_position;
  self.visited := init_visited;
end;

procedure santa.follow(const direction: char);
var 
  i: integer;
begin
  case direction of
    '^': self.position[0] := self.position[0] - 1;
    '>': self.position[1] := self.position[1] + 1;
    'v': self.position[0] := self.position[0] + 1;
    '<': self.position[1] := self.position[1] - 1;
  end;
  for i := low(self.visited^) to high(self.visited^) do begin
    if (self.position[0] = self.visited^[i][0]) and (self.position[1] = self.visited^[i][1]) then
      exit;
  end;
  setlength(self.visited^, length(self.visited^) + 1);
  self.visited^[high(self.visited^)] := self.position;
end;

const init_position: pair = (0, 0);
var
  input: file of char;
  symbol: char;
  bob: santa;
  visited: positions;
begin
  visited := [init_position];
  bob := santa.init(init_position, @visited);
  assign(input, 'input.txt');
  reset(input);
  while not eof(input) do begin
    read(input, symbol);
    bob.follow(symbol);
  end;
  close(input);
  writeln(length(visited));
end.
